### PR TITLE
Feature/expanded survey results

### PR DESCRIPTION
**Pull request by untra**
Added in a new endpoint: GET `api/results/<survey_id>`

This returns a formatted set of survey results, with embedded objects ready to be inserted into chartists bars and pies.

``` json
[  
   {  
      "bar_data":{  
         "series":[  
            [  
               0,
               0,
               1,
               1,
               0,
               1,
               1,
               1,
               0,
               0
            ]
         ],
         "labels":[  
            1,
            2,
            3,
            4,
            5,
            6,
            7,
            8,
            9,
            10
         ]
      },
      "pie_data":[  

      ],
      "options":[  

      ],
      "title":"rating",
      "id":"3c5c245c-9d5e-48b0-b79d-5e69003ebe5e",
      "results":[  
         4,
         8,
         3,
         6,
         7
      ],
      "total_responses":5,
      "response_format":"rating"
   },
   {  
      "bar_data":{  
         "series":[  
            0
         ],
         "labels":[  
            "alpha",
            "delta",
            "gamma"
         ]
      },
      "pie_data":[  
         {  
            "value":2,
            "name":"alpha"
         },
         {  
            "value":1,
            "name":"delta"
         },
         {  
            "value":2,
            "name":"gamma"
         }
      ],
      "options":[  
         "alpha",
         "beta",
         "gamma",
         "delta"
      ],
      "title":"multiple choice",
      "id":"60b534dc-77f5-4964-b2b5-1fd0f69b53f0",
      "results":[  
         "alpha",
         "gamma",
         "alpha",
         "gamma",
         "delta"
      ],
      "total_responses":5,
      "response_format":"multipleChoice"
   },
   {  
      "bar_data":[  

      ],
      "pie_data":[  

      ],
      "options":[  

      ],
      "title":"free response",
      "id":"81f3f49a-0fe7-4571-884c-056f488614e0",
      "results":[  
         "sed existimo te, sicut nostrum Triarium, minus ab eo delectari, quod ista Platonis, Aristoteli, Theophrasti orationis ornamenta neglexerit.",
         "Non eram nescius, Brute, cum, quae summis ingeniis exquisitaque doctrina philosophi Graeco sermone tractavissent, ea Latinis litteris mandaremus, fore ut hic noster labor in varias reprehensiones incurreret.",
         "quis enim tam inimicus paene nomini Romano est, qui Ennii Medeam aut Antiopam Pacuvii spernat aut reiciat, quod se isdem Euripidis fabulis delectari dicat, Latinas litteras oderit?",
         "Scaevolam M.que Manilium, ab iisque M.",
         "Ego autem quem timeam lectorem, cum ad te ne Graecis quidem cedentem in philosophia audeam scribere? quamquam a te ipso id quidem facio provocatus gratissimo mihi libro, quem ad me de virtute misisti."
      ],
      "total_responses":5,
      "response_format":"freeResponse"
   },
   {  
      "bar_data":[  

      ],
      "pie_data":[  
         {  
            "value":5,
            "name":"false"
         }
      ],
      "options":[  
         true,
         false
      ],
      "title":"true or false",
      "id":"ba692758-148c-4339-b26b-6d9653eb20ac",
      "results":[  
         false,
         false,
         false,
         false,
         false
      ],
      "total_responses":5,
      "response_format":"trueOrFalse"
   }
]
```

### CODE REVIEW

File: models/survey.py [hunk 1]
```diff
@@ -143,8 +143,52 @@ def mark_deleted(self, survey_id):
...
             return query
-        except err:
+        except Exception as err:
+            logging.error(err)
+            return []
+
+    def get_formatted_results(self, survey_id):
+        results = self.get_results(survey_id)
+        try:
+            query = r.expr(results).coerce_to('array').map(
```
Comment (nicot) [inline models/survey.py]: Is this query or the result of a query?

File: models/survey.py [hunk 2]
```diff
@@ -143,8 +143,52 @@ def mark_deleted(self, survey_id):
...
+                lambda question: r.db(self.DB).table('Question').get(question[0]).merge(r.expr({
+                    'results': question[1],
+                    'total_responses': question[1].count(),
+                    'pie_data': r.db(self.DB).table('Question').get(question[0]).do(
+                        lambda question_data: r.branch(
```
Comment (nicot) [inline models/survey.py]: Can you name these lambdas to improve the readability of this code?

File: models/survey.py [hunk 3]
```diff
@@ -143,8 +143,52 @@ def mark_deleted(self, survey_id):
...
+                            r.branch(
+                                (question_data['response_format'] == Question().RESPONSE_MULTIPLE_CHOICE),
+                                {
+                                    'labels': question[1].distinct(),
+                                    'series': [question[1].distinct().do(lambda val: question[1].filter(lambda foo: foo == val).count())]
+                                },
+                                (question_data['response_format'] == Question().RESPONSE_RATING),
+                                {
+                                    'labels': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10],
+                                    'series': [r.expr([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]).map(lambda val: question[1].filter(lambda foo: foo == val).count())]
```
Comment (nicot) [inline models/survey.py]: Would a range statement work here?

File: models/survey.py [hunk 4]
```diff
@@ -143,8 +143,52 @@ def mark_deleted(self, survey_id):
...
+    def get_formatted_results(self, survey_id):
+        results = self.get_results(survey_id)
+        try:
+            query = r.expr(results).coerce_to('array').map(
+                lambda question: r.db(self.DB).table('Question').get(question[0]).merge(r.expr({
+                    'results': question[1],
+                    'total_responses': question[1].count(),
+                    'pie_data': r.db(self.DB).table('Question').get(question[0]).do(
+                        lambda question_data: r.branch(
+                            (r.expr(question_data['response_format'] == Question().RESPONSE_MULTIPLE_CHOICE) | (question_data['response_format'] == Question().RESPONSE_TRUE_OR_FALSE)),
```
Comment (nicot) [inline models/survey.py]: Can you make these lines shorter? Python standard is 79 chars: https://www.python.org/dev/peps/pep-0008/#maximum-line-length

File: test/test_survey_results.py [hunk 5]
```diff
@@ -53,8 +54,8 @@ class TestSurveyResults(BaseAsyncTest):
     survey_data = {}
 
     def setUpClass():
-        logging.disable(logging.CRITICAL)
-        user_ids = [User().create_generic_item() for i in range(5)]
+        # logging.disable(logging.CRITICAL)
```
Comment (untra) [inline test/test_survey_results.py]: Gotta reenable this. whoops

File: test/test_survey_results.py [hunk 6]
```diff
@@ -80,11 +81,67 @@ def setUpClass():
...
+    def _test_formatted_response_free(self, formatted_result_data):
+        self.assertEqual(formatted_result_data['bar_data'], [])
+        self.assertEqual(formatted_result_data['pie_data'], [])
+        return
+
+    def _test_formatted_response_rating(self, formatted_result_data):
+        bar_data = formatted_result_data['bar_data']
+        series_data = bar_data['series'][0]
+        self.assertNotEqual(bar_data, [])
+        self.assertEqual(sorted(bar_data['labels']), sorted([1, 2, 3, 4, 5, 6, 7, 8, 9, 10]))
```
Comment (nicot) [inline test/test_survey_results.py]: Why are you sorting the second element?

File: test/test_survey_results.py [hunk 7]
```diff
@@ -80,11 +81,67 @@ def setUpClass():
...
+            "length of {0} does not match length of {1}".format(bar_data['labels'], series_data))
+        self.assertEqual(formatted_result_data['pie_data'], [])
+        return
+
+    def _test_formatted_response_true_or_false(self, formatted_result_data):
+        self.assertEqual(formatted_result_data['bar_data'], [])
+        self.assertNotEqual(formatted_result_data['pie_data'], [])
+        return
+
+    def _test_formatted_response_multiple_choice(self, formatted_result_data):
```
Comment (nicot) [inline test/test_survey_results.py:142]: So multiple choice sends back pie data and bar data? I think a bar chart would make more sense, as there will often be more than 100% response on multiple choice.

File: models/survey.py (hunk 2 above)
Comment (untra) [inline models/survey.py]: They are named. That one is called `question_data`

File: models/survey.py (hunk 1 above)
Comment (untra) [inline models/survey.py]: The result of the query. I'll change the name.

File: models/survey.py (hunk 3 above)
Comment (untra) [inline models/survey.py]: No, interestingly. I did think about that, lol.
I did try the range statements, but the rethink queries failed when I did.

File: test/test_survey_results.py (hunk 7 above)
Comment (untra) [inline test/test_survey_results.py:142]: Uhh this was discussed with @wannabeCitizen on monday. True/False are pies, Ratings are Bars, multiple choice is both.

File: models/survey.py (hunk 2 above)
Comment (nicot) [inline models/survey.py]: Oh thanks, I thought that question_data was the parameter.

File: test/test_survey_results.py (hunk 7 above)
Comment (nicot) [inline test/test_survey_results.py:142]: Sorry I missed that conversation.

How does a pie chart make sense if somebody votes for more than one answer? Don't pie charts always add up to 100?

Comment (untra) [inline test/test_survey_results.py:142]: Multiple choice is still only pick 1. There is no double voting on multiple chocie questions

**Comment by untra** [comment]
@MCChung303 this should be pulled in as soon as possble so you can get to work on making the graphs and visualizations for the frontend.

**Pull request merged**